File: tests/test-new-hooks.mjs
import assert from "node:assert/strict";
import { existsSync, mkdtempSync, readFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { describe, it } from "node:test";
import { parseHookOutput, runHook } from "./helpers.mjs";

describe("UserPromptSubmit", () => {
	it("should run without error", () => {
		const result = runHook("user-prompt-submit.mjs", { prompt: "hello" });
		assert.equal(result.status, 0);
	});

	it("should pass empty prompt", () => {
		const result = runHook("user-prompt-submit.mjs", { prompt: "" });
		assert.equal(result.status, 0);
	});
});

describe("PostFailure", () => {
	it("should run without error", () => {
		const result = runHook("post-failure.mjs", {
			tool_name: "Bash",
			tool_error: "command not found",
		});
		assert.equal(result.status, 0);
	});

	it("should log failure", () => {
		const result = runHook("post-failure.mjs", {
			tool_name: "Bash",
			tool_error: "permission denied",
		});
		assert.equal(result.status, 0);
	});
});

describe("SessionEnd", () => {
	it("should run without error", () => {
		const result = runHook("session-end.mjs", {});
		assert.equal(result.status, 0);
	});
});

describe("TeammateIdle", () => {
	it("should run and warn", () => {
		const result = runHook("teammate-idle.mjs", { agent_name: "hermes" });
		const output = parseHookOutput(result);
		assert.equal(result.status, 0);
		if (output?.hookSpecificOutput) {
			const ctx = output.hookSpecificOutput.additionalContext || "";
			assert.ok(
				ctx.toLowerCase().includes("hermes") ||
					ctx.toLowerCase().includes("idle"),
			);
		}
	});
});

describe("Notification", () => {
	it("should run without error", () => {
		const result = runHook("notification.mjs", {
			message: "test notification",
		});
		assert.equal(result.status, 0);
	});
});

describe("PermissionRequest", () => {
	it("should run without error", () => {
		const result = runHook("permission-request.mjs", {
			tool_name: "Bash",
			permission: "execute",
		});
		assert.equal(result.status, 0);
	});
});

describe("AuditLogging", () => {
	it("should write audit log when enabled", () => {
		const tmpDir = mkdtempSync(join(tmpdir(), "cca-audit-"));
		const result = runHook(
			"notification.mjs",
			{ message: "test" },
			{ CCA_HOOK_LOG_DIR: tmpDir },
		);
		assert.equal(result.status, 0);
		const logFile = join(tmpDir, "cca-hooks.jsonl");
		const lines = readFileSync(logFile, "utf8").trim().split("\n");
		assert.ok(lines.length >= 1, "Audit log should contain at least one entry");
		const entry = JSON.parse(lines[0]);
		assert.equal(entry.event, "Notification");
		assert.equal(entry.hook, "notification.mjs");
	});

	it("should not write log when disabled", () => {
		const tmpDir = mkdtempSync(join(tmpdir(), "cca-noaudit-"));
		const result = runHook("notification.mjs", { message: "test" });
		assert.equal(result.status, 0);
		const logFile = join(tmpDir, "cca-hooks.jsonl");
		assert.ok(!existsSync(logFile));
	});
});

File: tests/test_new_hooks.py
"""Tests for new hook scripts added in the hardening phase."""

import json
import os
import tempfile
from conftest import run_hook, parse_hook_output


class TestUserPromptSubmit:
    def test_runs_without_error(self):
        result = run_hook("user-prompt-submit.py", {"prompt": "hello"})
        assert result.returncode == 0

    def test_empty_prompt_passes(self):
        result = run_hook("user-prompt-submit.py", {"prompt": ""})
        assert result.returncode == 0


class TestPostFailure:
    def test_runs_without_error(self):
        result = run_hook("post-failure.py", {
            "tool_name": "Bash",
            "tool_error": "command not found",
        })
        assert result.returncode == 0

    def test_logs_failure(self):
        result = run_hook("post-failure.py", {
            "tool_name": "Bash",
            "tool_error": "permission denied",
        })
        assert result.returncode == 0


class TestSessionEnd:
    def test_runs_without_error(self):
        result = run_hook("session-end.py", {})
        assert result.returncode == 0


class TestTeammateIdle:
    def test_runs_and_warns(self):
        result = run_hook("teammate-idle.py", {"agent_name": "hermes"})
        output = parse_hook_output(result)
        assert result.returncode == 0
        if output:
            ctx = output.get("hookSpecificOutput", {}).get("additionalContext", "")
            assert "hermes" in ctx.lower() or "idle" in ctx.lower()


class TestNotification:
    def test_runs_without_error(self):
        result = run_hook("notification.py", {"message": "test notification"})
        assert result.returncode == 0


class TestPermissionRequest:
    def test_runs_without_error(self):
        result = run_hook("permission-request.py", {
            "tool_name": "Bash",
            "permission": "execute",
        })
        assert result.returncode == 0


class TestAuditLogging:
    def test_writes_audit_log_when_enabled(self):
        with tempfile.TemporaryDirectory() as tmpdir:
            result = run_hook("notification.py", {"message": "test"}, env={
                "CCA_HOOK_LOG_DIR": tmpdir,
            })
            assert result.returncode == 0
            log_file = os.path.join(tmpdir, "cca-hooks.jsonl")
            assert os.path.exists(log_file), "Audit log file should be created when CCA_HOOK_LOG_DIR is set"
            with open(log_file) as f:
                lines = f.readlines()
            assert len(lines) >= 1, "Audit log should contain at least one entry"
            entry = json.loads(lines[0])
            assert entry["event"] == "Notification"
            assert entry["hook"] == "notification.py"

    def test_no_log_when_disabled(self):
        with tempfile.TemporaryDirectory() as tmpdir:
            result = run_hook("notification.py", {"message": "test"})
            assert result.returncode == 0
            log_file = os.path.join(tmpdir, "cca-hooks.jsonl")
            assert not os.path.exists(log_file)
